### PR TITLE
Torch based OAC

### PR DESCRIPTION
**Pull request by ShuaibinLi**
### Add files:
mujoco_agent.py, mujoco_model.py, train.py, README.md, ---> benchmark/torch/
oac.py ---> parl/algorithms/torch/
### Change file:
parl/algorithms/torch/init.py

### CODE REVIEW

File: benchmark/torch/oac/train.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    parser.add_argument(
+        "--seed",
+        default=0,
+        type=int,
+        help='Sets Gym, PyTorch and Numpy seeds')
+    parser.add_argument(
+        "--start_timesteps",
+        default=1e4,
+        type=int,
+        help='Time steps initial exploration policy is used')
```
Comment (rical730) [inline benchmark/torch/oac/train.py]: Time steps initial, exploration policy is used

File: benchmark/torch/oac/train.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    parser.add_argument(
+        "--policy_freq",
+        default=1,
+        type=int,
+        help='Frequency to train actor and update params')
+    parser.add_argument(
+        "--automatic_entropy_tuning",
+        default=False,
+        type=bool,
+        help='Automatically adjust α')
```
Comment (rical730) [inline benchmark/torch/oac/train.py]: Whether or not to adjust alpha automatically

File: parl/algorithms/torch/oac.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                 alpha=None,
+                 beta=None,
+                 delta=None,
+                 actor_lr=None,
+                 critic_lr=None,
+                 decay=None,
+                 policy_freq=1,
+                 automatic_entropy_tuning=False,
+                 entropy_lr=None,
+                 action_space=None):
```
Comment (rical730) [inline parl/algorithms/torch/oac.py]: action_space？or action_dim?

File: parl/algorithms/torch/oac.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        assert isinstance(discount, float)
+        assert isinstance(tau, float)
+        assert isinstance(alpha, float)
+        assert isinstance(actor_lr, float)
+        assert isinstance(critic_lr, float)
+        assert isinstance(entropy_lr, float)
```
Comment (rical730) [inline parl/algorithms/torch/oac.py]: Is `entropy_lr` a necessary parameter or not ? if not, please remove this assert line.

Please check all the necessary parameter, and do the same check in SAC.

File: parl/algorithms/torch/oac.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        normal = Normal(act_mean, act_log_std.exp())
+        # for reparameterization trick  (mean + std*N(0,1))
+        x_t = normal.rsample()
+        y_t = torch.tanh(x_t)
+        action = y_t * self.max_action
+
+        log_prob = normal.log_prob(x_t)
+        log_prob -= torch.log(self.max_action * (1 - y_t.pow(2)) + 1e-6)
+        log_prob = log_prob.sum(1, keepdims=True)
+        # log_prob = torch.squeeze(log_prob, dim=1)
```
Comment (rical730) [inline parl/algorithms/torch/oac.py]: please remove this comment line.

File: parl/algorithms/torch/oac.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Get upper bound of Q
+        q1, q2 = self.model.critic_model(obs, tanh_mean)
+        mu_Q = (q1 + q2) / 2.0
+        sigma_Q = torch.abs(q1 - q2) / 2.0
+
+        # Upper Bound Q
+        Q_UB = mu_Q + self.beta * sigma_Q
+
+        # Obtain the gradient of Q_UB wrt to a with a evaluated at mu_t
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:103]: Can you comment this part of the code more generally？

Comment (ShuaibinLi) [reply parl/algorithms/torch/oac.py:103]: done

File: parl/algorithms/torch/oac.py (hunk 5 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/oac.py]: done

File: parl/algorithms/torch/oac.py (hunk 4 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/oac.py]: done

File: parl/algorithms/torch/oac.py (hunk 3 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/oac.py]: done

File: benchmark/torch/oac/train.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--env", default="Humanoid-v1", help='Mujoco gym environment name')
+    parser.add_argument(
+        "--seed",
+        default=0,
+        type=int,
+        help='Sets Gym, PyTorch and Numpy seeds')
+    parser.add_argument(
+        "--max_timesteps",
```
Comment (rical730) [inline benchmark/torch/oac/train.py]: please replace `max_timesteps ` by `train_total_steps`, and add `test_every_steps ` in  add_argument, refer https://github.com/PaddlePaddle/PARL/blob/develop/examples/SAC/train.py#L139

File: benchmark/torch/oac/mujoco_model.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+import parl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# clamp bounds for Std of action_log
+LOG_SIG_MAX = 2.0
+LOG_SIG_MIN = -20.0
+
```
Comment (rical730) [inline benchmark/torch/oac/mujoco_model.py:23]: please Initialize model weights

```
def weights_init_(m):
    if isinstance(m, nn.Linear):
        torch.nn.init.xavier_uniform_(m.weight, gain=1)
        torch.nn.init.constant_(m.bias, 0)
```

File: parl/algorithms/torch/oac.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        dist = Normal(mu_E, act_std)
+        z = dist.sample().detach()
+        action = torch.tanh(z) * self.max_action
+        return action
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        self._critic_learn(obs, action, reward, next_obs, terminal)
+        self._actor_learn(obs)
+
+        self.sync_target()
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:136]: return critic_cost, actor_cost

File: parl/algorithms/torch/oac.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
+        cur_q1, cur_q2 = self.model.critic_model(obs, action)
+
+        critic_loss = F.mse_loss(cur_q1, target_Q) + F.mse_loss(
+            cur_q2, target_Q)
+
+        self.critic_optimizer.zero_grad()
+        critic_loss.backward()
+        self.critic_optimizer.step()
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:153]: return critic_loss

File: parl/algorithms/torch/oac.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+    def _actor_learn(self, obs):
+        act, log_pi = self.predict(obs)
+        q1_pi, q2_pi = self.model.critic_model(obs, act)
+        min_q_pi = torch.min(q1_pi, q2_pi)
+        actor_loss = ((self.alpha * log_pi) - min_q_pi).mean()
+
+        self.actor_optimizer.zero_grad()
+        actor_loss.backward()
+        self.actor_optimizer.step()
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:164]: return actor_loss

File: parl/algorithms/torch/oac.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        with torch.no_grad():
+            next_action, next_log_pro = self.sample(next_obs)
+            q1_next, q2_next = self.target_model.critic_model(
+                next_obs, next_action)
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
+        cur_q1, cur_q2 = self.model.critic_model(obs, action)
+
+        critic_loss = F.mse_loss(cur_q1, target_Q) + F.mse_loss(
+            cur_q2, target_Q)
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:136]: critic_loss = critic_loss.mean()

File: parl/algorithms/torch/oac.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+class OAC(parl.Algorithm):
+    def __init__(self,
+                 model,
+                 gamma=None,
+                 tau=None,
+                 alpha=None,
+                 beta=None,
+                 delta=None,
+                 actor_lr=None,
+                 critic_lr=None):
```
Comment (rical730) [inline parl/algorithms/torch/oac.py:36]: please add Args description here.

File: parl/algorithms/torch/oac.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+        self.sync_target()
+
+    def _critic_learn(self, obs, action, reward, next_obs, terminal):
+        with torch.no_grad():
+            next_action, next_log_pro = self.sample(next_obs)
+            q1_next, q2_next = self.target_model.critic_model(
+                next_obs, next_action)
+            target_Q = torch.min(q1_next, q2_next) - self.alpha * next_log_pro
+            target_Q = reward + self.gamma * terminal * target_Q
```
Comment (rical730) [inline parl/algorithms/torch/oac.py]: (1.0 - terminal)

File: benchmark/torch/oac/train.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        # Select action use optimistic exploration or according to policy
+        if rpm.size() < WARMUP_STEPS:
+            action = np.random.uniform(-1, 1, size=action_dim)
+        else:
+            action = agent.sample(obs)
+
+        # Perform action
+        next_obs, reward, done, _ = env.step(action)
+        terminal = float(done) if episode_steps < env._max_episode_steps else 0
+        terminal = 1. - terminal
```
Comment (rical730) [inline benchmark/torch/oac/train.py]: please remove this line. terminal=False before episode finished

File: benchmark/torch/oac/train.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+from mujoco_agent import MujocoAgent
+from parl.algorithms import OAC
+
+WARMUP_STEPS = 1e4
+EVAL_EPISODES = 5
+MEMORY_SIZE = int(1e6)
+BATCH_SIZE = 256
+GAMMA = 0.99
+TAU = 0.005
+ALPHA = 0.2  # determines the relative importance of entropy term against the reward
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/oac/train.py]: Please put the `alpha`, `beta` and `delta` in the arguments.

Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/oac/train.py]: These arguments are unique to the OAC algorithm.

**Pull request closed without merging**